### PR TITLE
Add Conformance Test and implement L1 rule

### LINKED ISSUE #1: Add conformance tests
The Unicode consortium has a suite of conformance tests for the bidirectional algorithm, which can be found at:
- http://www.unicode.org/Public/UNIDATA/BidiTest.txt
- http://www.unicode.org/Public/UNIDATA/BidiCharacterTest.txt

We should add code to `tools/generate.py` to download these files and generate Rust unit tests for each of the test cases here.

### LINKED ISSUE #2: Implement L1: Adjust levels of whitespace characters
The `visual_runs` function does not yet implement [step L1](http://www.unicode.org/reports/tr9/#L1), which defines steps for adjusting the levels of certain whitespace and formatting characters.

In Gecko this is performed in [`nsBidi::AdjustWSLevels`](https://hg.mozilla.org/mozilla-central/file/7d4ab4a9febd/layout/base/nsBidi.cpp#l1286) and [`nsBidi::SetTrailingWSStart`](https://hg.mozilla.org/mozilla-central/file/7d4ab4a9febd/layout/base/nsBidi.cpp#l1494).

This may also be a good place to adjust levels for `removed_by_x9` characters, as in `nsBidi::AdjustWSLevels`.

### PR DESCRIPTION
**Pull request by behnam**
## Add conformance tests using BidiTest.txt
Initial results: `60494 test cases failed! (196253 passed)`

Because of the current limitations of rust's test framework, and the huge number of failures, the base (not including matching pairs) conformance test is executed in one run, and a summary is panic'ed (if there's any failure) and for the integration test to pass, it's marked as `should_panic`, with summary of the test run as `expected` string.

Fix #1 

## Implement L1 rule
To be able to implement L1, we need access to more information from
`BidiInfo`, namely `original_classes` of the `text`, in `visual_runs()`,
which would mean it should pass through `reorder_line()`.

The fact that information from `BidiInfo` is needed for both steps of
the public API (generating `BidiInfo` and consuming it
per-paragraph/per-level) made me change the API design and move these
methods into `impl BidiInfo`.

Then, since we needed access to `text` for every `BidiInfo` consumption,
I added a reference to `text` to `BidiInfo`, which also enables more
compile-time checks for `BidiInfo` isntance not outliving the text in
the user code.

NOTE: We are already breaking API in version 0.3.0 and paving for full
spec support is a good reason to do so, IMHO.

The L1 rule works by one pass on the text of the line.

Conformance Test: this implementation reduces the number of failures
from 60494 to 23770 (out of total 256747 cases).

Fix #2

---
This change is [Reviewable](https://reviewable.io/reviews/servo/unicode-bidi/30)

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -14,340 +14,428 @@
...
+        &self,
+        para: &ParagraphInfo,
+        line: Range<usize>,
+    ) -> (Vec<Level>, Vec<LevelRun>) {
+        assert!(line.start <= self.levels.len());
+        assert!(line.end <= self.levels.len());
+
+        let mut levels = self.levels.clone();
+
+        // Reset some whitespace chars to paragraph level.
```
Comment (behnam) [inline src/lib.rs]: Here's the L1 implementation.

**Comment by behnam** [comment]
Actually, found a bug in my L1 code that just fixed, updating the conf-test results: `12827 test cases failed! (243920 passed)`

**Comment by behnam** [comment]
Dependents status:
* https://github.com/servo/rust-url : works fine with this stack of changes, no updated needed.
* https://github.com/servo/servo : Depends on `unicode_bidi_serde`, which I'm working on now.

**Comment by mbrubeck** [comment]
@bors-servo r+

This is great, thanks!  Do you have any more changes planned before we publish version 0.3.0?

**Comment by behnam** [comment]
Thanks for the review,  @mbrubeck. Here's what I have in mind right now:
1. Make sure serde works and `servo` builds properly.
2. Working on the left over rules and try to reach 100% conformance for the base test.
3. Then add support for paired brackets tests and the rules.

I think I need a little help with the serde support, as looks like the current servo repo depends on serde `0.9`. I have a logically backward-compatible serde `1.0` implementation with tests right now (working on the `0.9` version this week), but then I realized that we may not need special implementation and `derive(Serialize, Deserialize)` could actually be enough.

So, the question is, are there any cases where a serde data from `unicode-bidi < 0.3` would be passed on to a serde deserializer of `unicode-bidi >= 0.3`?

Right now, I'm organizing serde implementation as two features: `with_serde0` and `with_serde1`, but they won't be able to be enabled at the same time. So, do we need to be able to build `unicode-bidi` with support for both `~1.0` and `<1.0` at the same time?

Also, I can see that https://github.com/serde-rs/legacy and relevant crates enable support for multiple versions of serde being linked, but I'm not sure how to do a similar thing for the serde_derive macros.

So, in short: if we only need either `with_serde0` or `with_serde1`, and there's no need for cross-`unicode-bidi`-version compatibility, I can condition them into the library with short and easy `derive` attributes. Otherwise, I would need more details on the expectation.

**Comment by mbrubeck** [comment]
> So, the question is, are there any cases where a serde data from unicode-bidi < 0.3 would be passed on to a serde deserializer of unicode-bidi >= 0.3?

No, I don't think we need to support this.
